### PR TITLE
Commands 

### LINKED ISSUE #79: Create Command-like interface
Information to get insight:
https://gameprogrammingpatterns.com/command.html

### LINKED ISSUE #210: Chat Commands
**Is your feature request related to a problem? Please describe.**
It's not.

**Describe the solution you'd like**
A way to add our chat commands, if possible with a simple permission system so we can give our friends permissions to spawn vehicles, weapons etc but not strangers. If someone without permissions would use a command you'd be able to automatically send a customizable message on chat such as (You don't have permission to spawn, please ask me nicely).

Also a cool idea using this system would be chat minigames, such as posting a small math operation and whoever answers the right answer, gets a lil reward or something.

**Provide Reasoning**
Would be cool for when we play with friends.

**Additional context**
None.

### LINKED ISSUE #734: [Bug]: No NPCs on the streets
### Describe the bug

No NPCs on the streets

### Steps To Reproduce

1. inject

### Expected Behavior

No bots walk the streets, only drive cars. To remove this bug, I deleted the hook hooks::create_script_handler
i also removed the code 
`for (auto& thread : *m_script_threads)
		{
			if (thread->m_handler)
				hook_script_handler((CGameScriptHandler*)thread->m_handler);
		}`
and removed natives hooks: NETWORK_SET_THIS_SCRIPT_IS_NETWORK_SCRIPT and NETWORK_TRY_TO_SET_THIS_SCRIPT_IS_NETWORK_SCRIPT.

After these actions, NPCs appeared on the streets. I'm not sure exactly which function (hook) of the above affects this

### Screenshots and/or Logs

_No response_

### Store

Rockstar

### Additional context

_No response_

### Build

- [X] I've confirmed the issue exists on the latest version of YimMenu

### LINKED ISSUE #735: [Bug]: Delete Gun, Gravity Gun, Repair Gun ( All the Gun mods ) stop working after a certain event that I can't pin point.
### Describe the bug

Happens every lobby that I manage to bug the options to use either of the Delete Gun, Repair Gun, Gravitiy Gun etc. It just doesn't shoot at all anymore. I can switch to " No Weapon Mod " And I can shoot normal bullets but no special bullet works either.

I think it happens everytime after you use the Delete Gun on a Player ( On accident )

### Steps To Reproduce

No clue, just play around with them

### Expected Behavior

-

### Screenshots and/or Logs

```
<paste your logs here>
```

### Store

Steam

### Additional context

_No response_

### Build

- [X] I've confirmed the issue exists on the latest version of YimMenu

### PR DESCRIPTION
**Pull request by maybegreat48**
* Add a command system (not all features are turned into commands yet)
* Commands can be used by other players through chat
* Block breakup kicks on others when host

Current design is based around classes that add themselves to a global map/array upon construction. Commands can be used anywhere in the menu (makes implementing things like reactions, etc easier) and is meant to be exposed to scripting languages, hotkeys and chat in the future

Closes #210
Closes #79
Fixes #734
Fixes #735

### CODE REVIEW

**Comment by tupoy-ya** [comment]
Would be nice to have something like source engine console where you can type commands dynamically and see logs.

Also just personal preference to use source's commands like `kill` instead of `suicide` to make learning this commands easier, i think FiveM does something like this.

Should i post command ideas here?

**Comment by maybegreat48** [comment]
I've never played a Source Engine game before, but you can absolutely give me some commands ideas

**Comment by tupoy-ya** [comment]
<details>
<summary>Console</summary>
Here is how console can look like:

image
Would be even better if we can get ImGui Viewports to work so we can drag console out of gta5 window.
</details>

<details>
<summary>Command ideas</summary>
Some descriptions are taken directly from source's  `help`  command

- `help`
Find help about a convar/concommand.
 
- `joaat`
Print joaat hash of a string in console or in chat if called from chat.

- `version`
Print git info in console or in chat if called from chat.

- `connect`
Connect to specified RID.

- `disconnect`
Leave current session and enter solo public lobby.

- `leave`
Leave current session and enter single player.

- `quit`
Closes the game immediately.

- `exit`
Closes the game immediately.
Alias of quit.

- `bind`
Bind a key.

- `alias`
Alias a command.

- `create_ped`
Creates specified ped near player.

- `create_vehicle`
Creates specified vehicle near player.

- `give_weapon` or maybe just `give`
Gives specified weapon to player.

- `god`
Toggle. Player becomes invulnerable.

- `cl_showoverlay`
When set to 1 will show something like this:
image
Can be devided into sub commands like `cl_showfps` and `cl_showpos`

- `cl_fov`
Change players FOV.

- `developer`
Set log level.
Ex. debug logs not showing.
Probably a bad idea as individual toggles are better.

</details>

**Comment by maybegreat48** [comment]
I'd prefer a drop-down/up console, something like the one Creation Engine uses. We're going to need an ImGui expert for that though, couldn't find any existing implementations

**Comment by tupoy-ya** [comment]
We can just use ImGui docking to be able to place it anywhere we want, the only problem with this is that it doesn't have animations.


https://user-images.githubusercontent.com/72797377/208714290-567a6522-7f11-40f5-84aa-f4fdc34c5f9f.mp4



Also i don't think YimMenu UI is suited for  drop-down/up console.

**Comment by Yimura** [comment]
Yeah the docker branch of ImGui is pretty neat

**Comment by TheGreenBandit** [comment]
> <details>
> 
> <summary>Console</summary>
> 
> Here is how console can look like:
> 
> 
> 
> image
> 
> Would be even better if we can get ImGui Viewports to work so we can drag console out of gta5 window.
> 
> </details>
> 
> 
> 
> <details>
> 
> <summary>Command ideas</summary>
> 
> Some descriptions are taken directly from source's  `help`  command
> 
> 
> 
> - `help`
> 
> Find help about a convar/concommand.
> 
>  
> 
> - `joaat`
> 
> Print joaat hash of a string in console or in chat if called from chat.
> 
> 
> 
> - `version`
> 
> Print git info in console or in chat if called from chat.
> 
> 
> 
> - `connect`
> 
> Connect to specified RID.
> 
> 
> 
> - `disconnect`
> 
> Leave current session and enter solo public lobby.
> 
> 
> 
> - `leave`
> 
> Leave current session and enter single player.
> 
> 
> 
> - `quit`
> 
> Closes the game immediately.
> 
> 
> 
> - `exit`
> 
> Closes the game immediately.
> 
> Alias of quit.
> 
> 
> 
> - `bind`
> 
> Bind a key.
> 
> 
> 
> - `alias`
> 
> Alias a command.
> 
> 
> 
> - `create_ped`
> 
> Creates specified ped near player.
> 
> 
> 
> - `create_vehicle`
> 
> Creates specified vehicle near player.
> 
> 
> 
> - `give_weapon` or maybe just `give`
> 
> Gives specified weapon to player.
> 
> 
> 
> - `god`
> 
> Toggle. Player becomes invulnerable.
> 
> 
> 
> - `cl_showoverlay`
> 
> When set to 1 will show something like this:
> 
> image
> 
> Can be devided into sub commands like `cl_showfps` and `cl_showpos`
> 
> 
> 
> - `cl_fov`
> 
> Change players FOV.
> 
> 
> 
> - `developer`
> 
> Set log level.
> 
> Ex. debug logs not showing.
> 
> Probably a bad idea as individual toggles are better.
> 
> 
> 
> </details>
> 
> 


We could probably put the console in its own tab, I'd have to make an example to really show it.

**Comment by pelecanidae** [comment]
I think this is a pretty good approach, but it's worth pointing out that you're doing half of the job of a scripting language here :p 

Embedding something like Wren or LuaJIT would give you a lot of the same things for free. True commands like `kick_to_interior` could stay their original C++ functions, with a small script binding added for each. E.g. `ScriptEvents.kick_to_interior(Player.from_name("foo"))`. Looped "commands" like no clip could stay their original implementation, with the no clip options exposed to the VM as globals or with setters/getters. E.g. `player.no_clip = true`. Then, the console is just a REPL to a VM instance; no manual command parsing needed. The "command context" can still exist as well, stored in some fields that get set before entering the VM.

A downside is that this complicates things along a different axis -- chat commands can no longer be as simple as `wanted 1`. Likewise, hotkeys and reactions become more difficult to do in-menu and instead require writing some scripts. Personally, this is a worthwhile tradeoff, but I completely understand why one might disagree.

Having said all that, nice design. The `looped_command` changes are very welcome. A lot of the `command` machinery is also good and might be simplifiable once scripting lands. Passing command names as string literals feels like it might start to smell in a few months though and is probably worth polishing soon. The weak typing/specification of arguments is another ugly part, but it's probably not worth investing the time in template magic to clean up.

**Comment by ghost** [comment]
Commands.. you mean like stand-commands? Press a key and type the name in and it does something?

**Comment by maybegreat48** [comment]
@pelecanidae Yeah I don't like using strings to identify commands either but I'm pretty sure it is the only way without having header files all over the place, at least not without redesigning the entire UI. Commands are meant to be used in multiple contexts like chat, hotkeys, a command palette, or even through a web service. I feel that this is the simplest way to support most such use-cases 

@FireFox101889 Somewhat, the "press the key" thing (whatever that is) is not implemented yet though

File: src/util/session.hpp [hunk 1]
```diff
@@ -152,28 +152,6 @@ namespace big::session
 	// TODO this is really broken
 	inline void enter_player_interior(player_ptr player)
 	{
```
Comment (BugisoftRSG) [inline src/util/session.hpp:154]: now a empty method just a note for later maybe

**Review by BugisoftRSG — APPROVED** [review]

**Comment by Yimura** [comment]
Merging this as-is, the framework is there and old features will continue to work.

The goal is to move other features to the command design over time.

**Pull request merged**